pkg/virt-handler/vm.go

```diff
@@ -656,6 +687,93 @@ func (d *VirtualMachineController) migrationTargetUpdateVMIStatus(vmi *v1.Virtua
 	return nil
 }
 
+func (d *VirtualMachineController) generateEventsForVolumeStatusChange(vmi *v1.VirtualMachineInstance, newStatusMap map[string]v1.VolumeStatus) {
+	for _, oldStatus := range vmi.Status.VolumeStatus {
+		newStatus, ok := newStatusMap[oldStatus.Name]
+		if !ok {
+			// status got removed
+			d.recorder.Event(vmi, k8sv1.EventTypeNormal, VolumeUnplugged, fmt.Sprintf("Volume %s has been unplugged", oldStatus.Name))
+			continue
+		}
+		if newStatus.Phase != oldStatus.Phase {
+			d.recorder.Event(vmi, k8sv1.EventTypeNormal, newStatus.Reason, newStatus.Message)
+		}
+		delete(newStatusMap, newStatus.Name)
+	}
+	// Send events for any new statuses.
+	for _, v := range newStatusMap {
+		d.recorder.Event(vmi, k8sv1.EventTypeNormal, v.Reason, v.Message)
+	}
```

Comment:
Thinking a head a bit. The newStatusMap is passed in by the caller function, and in this helper function items are deleted from it.

I don't know if there's any future use case where someone might want to use the newStatusMap, but i'd find it surprising if items were deleted by a helper function.

instead of deleting items, couldn't we create a new map in this helper which indicates what volume statuses have been handled, and then check that when iterating over the newStatusMap to determine what events are new?

Reply:
Good point, modified the function to make a copy and delete from there.